util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java

```diff
@@ -15,33 +15,62 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
-import java.util.function.Consumer;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
 import java.util.function.Supplier;
-
-final class ActorFutureCollector<V>
-    implements Consumer<ActorFuture<V>>, Supplier<ActorFuture<List<V>>> {
+import java.util.stream.Collector;
+
+/**
+ * Aggregates a number of {@code ActorFuture} objects into a single one. The aggregated future is
+ * completed when all individual futures have completed. If all futures complete with a value, the
+ * aggregated future returns the ordered list of said values. If one or more complete exceptionally,
+ * the aggregated future will complete exceptionally. The exception will have the exceptions of the
+ * individual futures aas suppressed exceptions. If exceptions occur, this does not interrupt the
+ * individual futures.
+ *
+ * <pre>
+ * var aggregated = of(future1, future2).stream().collect(new ActorFutureCollector<>(concurrencyControl));
+ * </pre>
+ *
+ * @param <V> type of the value of each future
+ */
+public final class ActorFutureCollector<V>
+    implements Collector<ActorFuture<V>, List<ActorFuture<V>>, ActorFuture<List<V>>> {
 
   private final ConcurrencyControl concurrencyControl;
-  private final List<ActorFuture<V>> futures = new ArrayList<>();
 
-  ActorFutureCollector(final ConcurrencyControl concurrencyControl) {
-    this.concurrencyControl = concurrencyControl;
+  public ActorFutureCollector(final ConcurrencyControl concurrencyControl) {
+    this.concurrencyControl = Objects.requireNonNull(concurrencyControl);
+  }
+
+  @Override
+  public Supplier<List<ActorFuture<V>>> supplier() {
+    return ArrayList::new;
   }
 
-  ActorFutureCollector(final ActorFutureCollector<V> a, final ActorFutureCollector<V> b) {
-    concurrencyControl = a.concurrencyControl;
-    a.futures.forEach(this);
-    b.futures.forEach(this);
+  @Override
+  public BiConsumer<List<ActorFuture<V>>, ActorFuture<V>> accumulator() {
+    return List::add;
+  }
+
+  @Override
+  public BinaryOperator<List<ActorFuture<V>>> combiner() {
+    return (listA, listB) -> {
+      listA.addAll(listB);
+      return listA;
+    };
   }
 
   @Override
-  public void accept(final ActorFuture<V> actorFuture) {
-    futures.add(Objects.requireNonNull(actorFuture));
+  public Function<List<ActorFuture<V>>, ActorFuture<List<V>>> finisher() {
+    return futures -> new CompletionWaiter<>(concurrencyControl, futures).get();
   }
 
   @Override
-  public ActorFuture<List<V>> get() {
-    return new CompletionWaiter<>(concurrencyControl, futures).get();
+  public Set<Characteristics> characteristics() {
+    return Collections.singleton(Characteristics.CONCURRENT);
```

Comment:
:question: I'm still confused by the characteristics :smile: Why did we add CONCURRENT now?

Reply:
Actually, this is wrong. Yesterday I didn't think hard about it. Today I thought I could set this characteristic because I thought it is about the combiner function, which works concurrently. The JavaDoc isn't too detailed but it references the accumulator, which does not work concurrently. So I removed it again.